core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java

```diff
@@ -662,7 +667,10 @@ private void addGatewayUserFacingCertAndFqdn(GatewayConfig gatewayConfig, Cluste
 
     private List<String> getRangerFqdn(Cluster cluster, String primaryGatewayFqdn, List<String> rangerLocations) {
         if (rangerLocations.size() > 1) {
-            String gatewayGroupName = cluster.getStack().getPrimaryGatewayInstance().getInstanceGroup().getGroupName();
+            Stack stack = cluster.getStack();
+            Set<InstanceGroup> groups = instanceGroupService.findByStackId(stack.getId());
+            stack.setInstanceGroups(groups);
```

Comment:
I don't like that we're modifying the Stack object here. I know you're just setting it with the instance groups that are part of the stack, but the Cluster and Stack objects are pass around a lot and I don't think we should modify them unless we have to.

There's a method in `InstanceMetaDataService` called `getPrimaryGatewayInstanceMetadata()`, which wraps to the SQL query to fetch that object. I think you should add a similar query to the `InstanceGroupService`/`InstanceGroupRepository` classes and just call that directly. Like:
```
InstanceGroup instanceGroup = instanceGroupService.getPrimaryGatewayInstanceGroup(cluster.getStack.getId());
if (instanceGroup != null) {
    String gatewayGroupName = instanceGroup.getGroupName();
    ....
```

Reply:
The stack object's getPrimaryGatwayinstance method does a lot of filtering of the instance groups and instance group metadata.  There's a lot of logic i don't want to reproduce here, so setting it gets it for free.  I can unset it afterwards if needed.

Reply:
The `stack.getPrimaryGatewayInstance()` is functionally just going through all the instance metadata looking for an instance group of type `GATEWAY_PRIMARY`. The `InstanceMetaDataService.getPrimaryGatewayInstanceMetadata()` does that with a single SQL query, and it looks like it actively loads the instance group in the repository class:
```
    @EntityGraph(value = "InstanceMetaData.instanceGroup", type = EntityGraph.EntityGraphType.LOAD)
    @Query("SELECT i FROM InstanceMetaData i WHERE i.instanceMetadataType = 'GATEWAY_PRIMARY' AND i.instanceStatus <> 'TERMINATED' "
            + "AND i.instanceGroup.stack.id= :stackId")
    Optional<InstanceMetaData> getPrimaryGatewayInstanceMetadata(@Param("stackId") Long stackId);
```
So I don't think you even need to modify the `InstanceGroupService`. I'm pretty sure you can just do this:
```
Optional<InstanceMetadata> instanceMetadata = instanceMetaDataRepository.getPrimaryGatewayInstanceMetadata(cluster.getStack().getId());
if (instanceMetata.isPresent) {
    String gatewayGroupName = instanceMetadata.get().getInstanceGroup().getGroupName();
    ...
}
```
And then you wouldn't be introducing new eager loading in `InstanceGroup` or modifying the `Stack` object.